cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,328 @@
+            self.mutation_priors = {}
+            probabilites = [
+                state_generating_distribution() for _ in range(number_of_states)
+            ]
+            Z = np.sum(probabilites)
+            for i in range(number_of_states):
+                self.mutation_priors[i + 1] = probabilites[i] / Z
+        else:
+            Z = np.sum([v for v in self.mutation_priors.values()])
+            if not math.isclose(Z, 1.0):
+                raise DataSimulatorError("Mutation priors do not sum to 1.")
+
+        self.heritable_silencing_rate = heritable_silencing_rate
+        self.stochastic_silencing_rate = stochastic_silencing_rate
+
+        self.random_seed = random_seed
+
+    def overlay_data(self, tree: CassiopeiaTree):
+        """Overlays Cas9-based lineage tracing data onto the CassiopeiaTree.
+
+        Args:
+            tree: Input CassiopeiaTree
+        """
+
+        if self.random_seed is not None:
+            np.random.seed(self.random_seed)
+
+        number_of_characters = self.number_of_cassettes * self.size_of_cassette
+
+        # initialize character states
+        character_matrix = {}
+        for node in tree.nodes:
+            character_matrix[node] = [-1] * number_of_characters
+
+        for node in tree.depth_first_traverse_nodes(tree.root, postorder=False):
+
+            if tree.is_root(node):
+                character_matrix[node] = [0] * number_of_characters
+                continue
+
+            parent = tree.parent(node)
+            t = tree.get_time(node) - tree.get_time(parent)
+
+            character_array = character_matrix[parent]
+            open_sites = [
+                c
+                for c in range(len(character_array))
+                if character_array[c] == 0
+            ]
+
+            new_cuts = []
+            for site in open_sites:
+                mutation_rate = self.mutation_rate_per_character[site]
+                p = 1 - (np.exp(-t * mutation_rate))
+
+                if np.random.uniform() < p:
+                    new_cuts.append(site)
+
+            # collapse cuts that are on the same cassette
+            cuts_remaining = new_cuts
+            if self.size_of_cassette > 1:
+                character_array, cuts_remaining = self.collapse_sites(
+                    character_array, new_cuts
+                )
+
+            # introduce new states at cut sites
+            character_array = self.introduce_states(
+                character_array, cuts_remaining
+            )
+
+            # silence cassettes
+            character_array = self.silence_cassettes(
+                character_array, self.heritable_silencing_rate
+            )
+
+            character_matrix[node] = character_array
+
+        # apply stochastic silencing
+        for leaf in tree.leaves:
+            character_matrix[leaf] = self.silence_cassettes(
+                character_matrix[leaf], self.stochastic_silencing_rate
+            )
```

Comment:
I think it would be *great* if we allowed the user to specify what integers to use to represent stochastic and heritable dropouts. That way, we can easily simulate data where we know whether a dropout in the character matrix is heritable or stochastic. This is really easy to implement: we just have to add two more arguments to `__init__`. I think as long as we require that they are negative numbers, and by default they are `-1`, we should be good. I know that some downstream methods assume that the missing state is `-1`, so we should definitely add a warning.

Reply:
That's a good idea! I'll add this to the simulator.